domain/port/state/state.go

```diff
@@ -552,20 +552,15 @@ WHERE uuid IN ($portRangeUUIDs[:])
 }
 
 // GetEndpoints returns all the endpoints for the given unit
-//
-// TODO(jack-w-shaw): At the moment, we calculate this by checking the unit_endpoints
-// table. However, this will not always return a complete list, as it only includes
-// endpoints that have had ports opened on them at some point.
-//
-// Once it has been implemented, we should check the charm_relation table to get a
-// complete list of endpoints instead.
 func (st *State) GetEndpoints(ctx domain.AtomicContext, unit coreunit.UUID) ([]string, error) {
 	unitUUID := unitUUID{UUID: unit}
 
 	getEndpoints, err := st.Prepare(`
-SELECT &endpointName.*
-FROM unit_endpoint
-WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
+SELECT charm_relation.key AS &endpointName.endpoint
+FROM unit
+JOIN application ON unit.application_uuid = application.uuid
+JOIN charm_relation ON application.charm_uuid = charm_relation.charm_uuid
+WHERE unit.uuid = $unitUUID.unit_uuid
```

Comment:
This is starting to look like this should be a view.

Reply:
Not convinced at the moment. This query is only made here atm